subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
             return false;
         }
     }
+
+    private void detectMissingDependencies(LocalTaskNode node) {
+        RelatedLocations consumedDirectories = consumedAndProducedLocations.getConsumedDirectories();
+        for (String outputPath : node.getMutationInfo().outputPaths) {
+            consumedDirectories.getNodesRelatedTo(outputPath).stream()
+                .filter(consumerNode -> missesDependency(node, consumerNode))
+                .forEach(consumerWithoutDependency -> emitMissingDependencyDeprecationWarning(node, consumerWithoutDependency));
+        }
+        Set<String> consumedLocations = new LinkedHashSet<>();
+        node.getTaskProperties().getInputFileProperties()
+            .forEach(spec -> spec.getPropertyFiles().visitStructure(new FileCollectionStructureVisitor() {
+                @Override
+                public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
+                    contents.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitGenericFileTree(FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    fileTree.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
+                    consumedLocations.add(root.getAbsolutePath());
+                }
+
+                @Override
+                public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    consumedLocations.add(file.getAbsolutePath());
+                }
+            }));
+        consumedDirectories.recordRelatedToNode(node, consumedLocations);
+        for (String consumedLocation : consumedLocations) {
+            consumedAndProducedLocations.getProducedDirectories().getNodesRelatedTo(consumedLocation).stream()
+                .filter(producerNode -> missesDependency(producerNode, node))
+                .forEach(producerWithoutDependency -> emitMissingDependencyDeprecationWarning(producerWithoutDependency, node));
+        }
+    }
+
+    private static boolean missesDependency(Node producer, Node consumer) {
+        if (consumer == producer) {
+            return false;
+        }
+        if (consumer.getDependencySuccessors().contains(producer)) {
+            return false;
+        }
+        for (Node dependency : consumer.getAllSuccessors()) {
+            if (dependency == producer || dependency.getDependencySuccessors().contains(producer)) {
+                return false;
+            }
+        }
```

Comment:
What's the difference between searching in `getDependencySuccessors()` vs `getAllSuccessors()`? It seems we should not be doing both.

Reply:
`getAllSuccessors` includes `mustRunAfter` and finalizers as well. `getDependencySuccessors()` only includes actual hard dependencies.

The reason why we use both methods here is a performance optimization: We can easily check if something is part of the `dependencySuccessors`, since it is a set. `allSuccessors` is only an `Iterable`, so we need to iterate over it to check whether a node is in there or not.

Reply:
Please add a comment about why we need this.

Reply:
Added.